settings/src/components/webauthn/list-keys.js

```diff
@@ -1,38 +1,160 @@
+			modalKey.delete_nonce,
+
+			// Success handler.
+			() => {
+				setDeleting( false );
+				setGlobalNotice( modalKey.name + ' has been deleted.' );
+				setModalKey( null );
+				refreshRecord( userRecord );
+			},
+
+			// Error handler.
+			( error ) => {
+				setModalError( error );
+				setDeleting( false );
+			}
+		);
+	}, [ modalKey ] );
+
+	return (
+		<>
+			<ul>
+				{ keys.map( ( key ) => (
+					<li key={ key.id }>
+						{ key.name }
+
+						<Button
+							variant="link"
+							data-id={ key.id }
+							aria-label="Delete"
+							onClick={ () => setModalKey( key ) }
+						>
+							Delete
+						</Button>
+					</li>
+				) ) }
+			</ul>
+
+			{ modalKey && (
+				<ConfirmRemoveKey
+					keyToRemove={ modalKey }
+					error={ modalError }
+					deleting={ deleting }
+					onClose={ () => setModalKey( null ) }
+					onConfirm={ onConfirmDelete }
+				/>
+			) }
+		</>
+	);
+}
+
+/**
+ * Prompt the user to confirm they want to delete a key.
  *
- * @param props
- * @param props.keys
+ * @param {Object}   props
+ * @param {Object}   props.keyToRemove
+ * @param {Function} props.onConfirm
+ * @param {Function} props.onClose
+ * @param {boolean}  props.deleting
+ * @param {string}   props.error
  */
-export default function ListKeys( { keys } ) {
+function ConfirmRemoveKey( { keyToRemove, onConfirm, onClose, deleting, error } ) {
 	return (
-		<ul>
-			{ keys.map( ( key ) => (
-				<li key={ key.id }>
-					{ key.name }
-
-					{ /* todo add onclick handler that pops up a <Modal> to confirm. maybe pass in from parent? */ }
-					<Button
-						variant="link"
-						data-id={ key.id }
-						aria-label="Delete"
-						onClick={ () =>
-							confirm(
-								'Modal H4 Remove Key? p Are you sure you want to remove the "" security key? Button Cancel Button Remove Key'
-							)
-						}
-					>
-						Delete
-					</Button>
-				</li>
-			) ) }
-		</ul>
+		<Modal
+			title={ `Remove ${ keyToRemove.name }?` }
+			className="wporg-2fa__confirm-delete-key"
+			onRequestClose={ onClose }
+		>
+			<p>
+				Are you sure you want to remove the <code>{ keyToRemove.name }</code> security key?
+			</p>
+
+			<div className="wporg-2fa__submit-actions">
+				<Button variant="primary" onClick={ onConfirm }>
+					Remove { keyToRemove.name }
+				</Button>
+
+				<Button variant="secondary" onClick={ onClose }>
+					Cancel
+				</Button>
+			</div>
+
+			{ deleting && (
+				<p>
+					<Spinner />
+				</p>
+			) }
+
+			{ error && (
+				<Notice status="error" isDismissible={ false }>
+					<Icon icon={ cancelCircleFilled } />
+					{ error }
+				</Notice>
+			) }
+		</Modal>
 	);
 }
+
+/**
+ * POST an AJAX request to remove a key.
+ *
+ * @param {number}   userId
+ * @param {string}   credentialId
+ * @param {string}   nonce
+ * @param {Function} onSuccess
+ * @param {Function} onError
+ */
+async function handleRemoveKey( userId, credentialId, nonce, onSuccess, onError ) {
+	try {
+		await ajax.post( 'webauthn_delete_key', {
```

Comment:
If you return the promise, you'll be able to use the default `then`, `catch`, `finally` and maybe not need this function.

`finally` is nice because you wouldn't need to make multiple calls to `setDeleting( false );`. You can call that once in `finally`.

Reply:
That's a good point, i moved the func into `ListKeys` in b36783afae4338a7b250e9a9789881e5b0ecf887 so that it'd have direct access to the component members, rather than having to pass them in 👍🏻 

I also moved `setDeleting` to a `try...finally` block. 👍🏻 

At that point, the original `onConfirmDelete` function wasn't doing much, so I combined it with `handleRemoveKey`. I like this iteration a lot better, thanks! 🎉 

I still prefer to use `await`, since promising chaining has always felt very clunky to me. `await` feels conceptually simpler, and also much easier to read.